R/scale-discrete-.r

```diff
@@ -141,35 +141,104 @@ ScaleDiscretePosition <- ggproto("ScaleDiscretePosition", ScaleDiscrete,
   }
 )
 
-# TODO: This is a clear candidate for vctrs once we adopt it
-new_mapped_discrete <- function(x) {
-  if (is.null(x)) {
-    return(x)
-  }
-  if (!is.numeric(x)) {
-    cli::cli_abort("{.cls mapped_discrete} objects can only be created from numeric vectors")
-  }
-  class(x) <- c("mapped_discrete", "numeric")
-  x
+new_mapped_discrete <- function(x = double()) {
+  vec_assert(x, double())
+  obj <- new_vctr(x, class = "ggplot2_mapped_discrete")
+  # vctrs does not support inheriting from numeric base class
+  class(obj) <- c(class(obj), "numeric")
+  obj
+}
+mapped_discrete <- function(x = double()) {
+  if (is.null(x)) return(NULL)
+  if (is.array(x)) x <- as.vector(x)
+  new_mapped_discrete(vec_cast(x, double()))
+}
+is_mapped_discrete <- function(x) inherits(x, "ggplot2_mapped_discrete")
+#' @export
+format.ggplot2_mapped_discrete <- function(x, ...) format(vec_data(x), ...)
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.ggplot2_mapped_discrete <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.numeric <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.numeric.ggplot2_mapped_discrete <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.double <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.double.ggplot2_mapped_discrete <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.integer <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.integer.ggplot2_mapped_discrete <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.character <- function(x, y, ...) character()
+#' @export
+vec_ptype2.character.ggplot2_mapped_discrete <- function(x, y, ...) character()
+#' @export
+vec_ptype2.ggplot2_mapped_discrete.factor <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_ptype2.factor.ggplot2_mapped_discrete <- function(x, y, ...) new_mapped_discrete()
+#' @export
+vec_cast.ggplot2_mapped_discrete.ggplot2_mapped_discrete <- function(x, to, ...) x
+#' @export
+vec_cast.ggplot2_mapped_discrete.numeric <- function(x, to, ...) new_mapped_discrete(x)
+#' @export
+vec_cast.numeric.ggplot2_mapped_discrete <- function(x, to, ...) vec_data(x)
```

Comment:
We tend to use this a lot vs `vec_data()` in cast methods like this one - if you decide to avoid using `vec_data()` it may be useful

```r
> vctrs:::unstructure
function (x) 
{
    attributes(x) <- NULL
    x
}
```

Reply:
oh do we? I don't remember this vctrs version but I see that it removes names dimensions and everything. Seeing it, I'd think of it as a purely internal tool that shouldn't be used by others. `rlang:::unstructure()` is a bit safer to use.

Reply:
Oh yea the rlang one looks good

Reply:
since mapped discrete are only constructed internally from bare numerics it should be fine to use `vec_data()`